src/cstringholder.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright 2018, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use std::ffi::CStr;
+use std::ops::Deref;
+use std::os::raw::c_char;
+use std::ptr;
+use translate::*;
+use types::{StaticType, Type};
+
+use gobject_ffi;
+use ffi as glib_ffi;
+use value::{FromValueOptional, Value};
+
+#[derive(Debug, Hash, PartialOrd, Ord)]
+pub struct CStringHolder {
+    slice: Box<CStr>,
+    ptr: *const c_char,
+    owned: bool,
+}
+
+impl CStringHolder {
+    pub fn new(ptr: *const c_char) -> Self {
+        assert!(!ptr.is_null());
+        unsafe {
+            let slice = CStr::from_ptr(ptr);
+            Self { ptr, slice: Box::from(slice), owned: true }
+        }
+    }
+
+    pub fn from(s: &[u8]) -> Self {
+        let slice = match CStr::from_bytes_with_nul(s) {
+            Ok(s) => s,
+            Err(e) => panic!(e),
+        };
+        let ptr = slice.as_ptr();
+        assert!(!ptr.is_null());
+        Self { ptr, slice: Box::from(slice), owned: false }
+    }
+
+    pub fn as_str(&self) -> &str {
+        let bytes = self.slice.to_bytes();
+        match std::str::from_utf8(bytes) {
+            Ok(s) => s,
+            Err(e) => {
+                panic!("UTF-8 conversion failed: {}", e);
+            },
+        }
+    }
+}
+
+impl Drop for CStringHolder {
+    fn drop(&mut self) {
+        if self.owned {
+            unsafe {
+                glib_ffi::g_free(self.ptr as *mut _);
+            }
+        }
+    }
+}
+
+impl PartialEq for CStringHolder {
+    fn eq(&self, other: &CStringHolder) -> bool {
+        self.slice == other.slice
+    }
+}
+
+impl Eq for CStringHolder {}
+
+impl AsRef<str> for CStringHolder {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl Deref for CStringHolder {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl From<CStringHolder> for String {
+    fn from(holder: CStringHolder) -> Self {
+        String::from(holder.as_str())
+    }
+}
+
+impl From<String> for CStringHolder {
```

Comment:
This one would involve copying, not sure where it would be useful to use?

Reply:
Yeah let's remove it, doesn't seem useful indeed :)